### PR TITLE
Design-proposal: KubeVirt DRA design proposal

### PR DESCRIPTION
**Pull request by rthallisey**
DRA ([Dynamic Resource Allocation](https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/)) design proposal.

#### What this PR does / why we need it:
KubeVirt DRA integration will allow VM users fine-grained control of devices.  This is important for many use-cases:
- Full Device Advertisement 
In the current device-plugin model, GPUs can only be advertised in one mode (Passthrough, MIG, vGPU) for the full lifecycle of the device.  This model doesn’t work when an end-user has the ability to choose which type of device they need - PT, vGPU, or MIG.
Advertising full devices is a common use-case in ephemeral workloads like HPC, ML, Cloud Gaming, ect.., that require on-demand driver switches on the GPU, based on the type of workload requested.
- Pluggable CPU Management
Similar to the GPU, the CPU can be logically separated and handed out to workloads.  Conceptually the use-case would be the same as GPUs - allow a plugin to handle the specialized allocation of the CPU.  Similarly, need memory management as well.  We should be allowing full platform management for specialized vertical solutions.

DRA is not a device-plugin replacement - it solves more problems.

#### Release note:
```release-note
None
```

### CODE REVIEW

**Comment by rthallisey** [comment]
cc @varunrsekar

File: design-proposals/dra.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  running: false
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      resourceClaims:
+      - name: a100-40C
+        source:
+          resourceClaimTemplateName: a100-40C-claim-template
```
Comment (varunrsekar) [inline design-proposals/dra.md:135]: Specifically for GPUs, this is the spec I had in mind to keep in-sync with the existing architecture and to address an usecase such as:
> As a user, I would like to use the dra driver for certain devices and device-plugins for others.

```
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm-cirros
  name: vm-cirros
spec:
  running: false
  template:
    metadata:
      labels:
        kubevirt.io/vm: vm-cirros
    spec:
       domain:
          devices:
             gpus:
             - name: gpu0 
               resourceClaim:
                  name: a100-40C
                  source:
                     resourceClaimTemplateName: a100-40C-claim-template
–--
apiVersion: v1
kind: Pod
metadata:
  name: virt-launcher-cirros
spec:
  containers:
  - name: virt-launcher
    image: virt-launcher
    resources:
      claims:
      - name: a100-40C
  resourceClaims:
  - name: a100-40C
    source:
      resourceClaimTemplateName: a100-40C-claim-template
```

The idea is that, for a `gpu` device, if the `resourceClaim` field is set, then kubevirt will know its to be provisioned by a dra driver and applies it to the virt-launcher pod spec. If instead, the `deviceName` is set, then kubevirt will know it is to be provisioned by a device-plugin and will set `resources.requests[<deviceName>]`.

The same rationale can be applied to the other device types from https://pkg.go.dev/kubevirt.io/api/core/v1#DomainSpec that would need DRA integration

**Comment by alicefr** [comment]
@rthallisey one thing that is difficult for me to model with DRA is how we pass the device information. Today, there is an implicit API/mechanism between the device plugin and kubevirt how the device information are passed through environmental variables. For example, kubevirt cannot work out-of-the-box with any device plugins. 
Here, we kind have the some problem. Do you have any ideas how we could formalize more the mechanism how to pass the device information?

File: design-proposals/dra.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  running: false
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      resourceClaims:
+      - name: rtx4090
+        source:
+          resourceClaimTemplateName: rtx4090-claim-template
```
Comment (alicefr) [inline design-proposals/dra.md:73]: Do we want to reference a claim here or the claim template? Can the user set the spec of the template?

Comment (fabiand) [reply design-proposals/dra.md:73]: We have to have both.

**Comment by rthallisey** [comment]
@alicefr 
>@rthallisey one thing that is difficult for me to model with DRA is how we pass the device information. Today, there is an implicit API/mechanism between the device plugin and kubevirt how the device information are passed through environmental variables. For example, kubevirt cannot work out-of-the-box with any device plugins.
Here, we kind have the some problem. Do you have any ideas how we could formalize more the mechanism how to pass the device information?
>
@varunrsekar is discussing this with upstream DRA folks.  We're working on a path to formalize device info in vendor plugins.

**Comment by aburdenthehand** [comment]
@rthallisey @varunrsekar Are we still going ahead with this?

**Comment by rthallisey** [comment]
@aburdenthehand yes

**Comment by rthallisey** [comment]
Discussion has moved to https://github.com/kubevirt/community/pull/331

**Pull request closed without merging**